src/backend/distributed/commands/table.c

```diff
@@ -502,6 +562,131 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
+ * to a citus local table.
+ */
+static void
+ErrorIfAlterTableDefinesFKeyFromPostgresToCitusLocalTable(
+	AlterTableStmt *alterTableStatement)
+{
+	List *commandList = alterTableStatement->cmds;
+
+	LOCKMODE lockmode = AlterTableGetLockLevel(commandList);
+	Oid relationId = AlterTableLookupRelation(alterTableStatement, lockmode);
+
+	if (IsCitusTable(relationId))
+	{
+		/* left relation is not a postgres local table, */
+		return;
+	}
+
+	List *alterTableFKeyConstraints =
+		GetAlterTableStmtFKeyConstraintList(alterTableStatement);
+	Constraint *constraint = NULL;
+	foreach_ptr(constraint, alterTableFKeyConstraints)
+	{
+		Oid rightRelationId = RangeVarGetRelid(constraint->pktable, lockmode,
+											   alterTableStatement->missing_ok);
+		if (IsCitusLocalTable(rightRelationId))
+		{
+			ErrorOutForFKeyBetweenPostgresAndCitusLocalTable(relationId);
+		}
+	}
+}
+
+
+/*
+ * GetAlterTableStmtFKeyConstraintList returns a list of Constraint objects for
+ * the foreign keys that given ALTER TABLE statement defines.
+ */
+static List *
+GetAlterTableStmtFKeyConstraintList(AlterTableStmt *alterTableStatement)
+{
+	List *alterTableFKeyConstraintList = NIL;
+
+	List *commandList = alterTableStatement->cmds;
+	AlterTableCmd *command = NULL;
+	foreach_ptr(command, commandList)
+	{
+		List *commandFKeyConstraintList = GetAlterTableCommandFKeyConstraintList(command);
+		alterTableFKeyConstraintList = list_concat(alterTableFKeyConstraintList,
+												   commandFKeyConstraintList);
+	}
+
+	return alterTableFKeyConstraintList;
+}
+
+
+/*
+ * GetAlterTableCommandFKeyConstraintList returns a list of Constraint objects
+ * for the foreign keys that given ALTER TABLE subcommand defines. Note that
+ * this is only possible if it is an:
+ *  - ADD CONSTRAINT subcommand (explicitly defines) or,
+ *  - ADD COLUMN subcommand (implicitly defines by adding a new column that
+ *    references to another table.
+ */
+static List *
+GetAlterTableCommandFKeyConstraintList(AlterTableCmd *command)
+{
+	List *fkeyConstraintList = NIL;
+
+	AlterTableType alterTableType = command->subtype;
+	if (alterTableType == AT_AddConstraint)
+	{
+		Constraint *constraint = (Constraint *) command->def;
+		if (constraint->contype == CONSTR_FOREIGN)
+		{
+			fkeyConstraintList = lappend(fkeyConstraintList, constraint);
+		}
+	}
+	else if (alterTableType == AT_AddColumn)
+	{
+		ColumnDef *columnDefinition = (ColumnDef *) command->def;
+		List *columnConstraints = columnDefinition->constraints;
+
+		Constraint *constraint = NULL;
+		foreach_ptr(constraint, columnConstraints)
+		{
+			if (constraint->contype == CONSTR_FOREIGN)
+			{
+				fkeyConstraintList = lappend(fkeyConstraintList, constraint);
+			}
+		}
+	}
+
+	return fkeyConstraintList;
+}
+
+
+/*
+ * AlterTableCommandTypeIsTrigger returns true if given alter table command type
+ * is identifies an ALTER TABLE .. TRIGGER .. command.
+ */
+static bool
+AlterTableCommandTypeIsTrigger(AlterTableType alterTableType)
```

Comment:
In general, we prefer function names like `XIsSomething` over `IsXSomething` by convention in code-base as far as I remember from one of our previous talks with @onderkalaci

Reply:
I am not sure if we have such a convention, I see many many usages of functions starting with `Is`. 

I think the prefix of a function is probably the most important part of its name, because it directly gives away what the function does. For example if a function starts with `Get` we easily have an idea of what kind of thing it is doing, if a function starts with `Set` we know it is setting some fields. Similar to that if a function starts with `Is` `Are` we know that it returns a boolean. 

Even if we have such a convention, I think we should considering changing it. Maybe there is some reason behind it but what better clarity does `XIsSomething` over `IsXSomething`?

Reply:
We chose to go with `XIsSomething` for now as we wanted to get input from @onderkalaci as well, so please comment on this thread @onderkalaci.